js/node_shared/unix.js

```diff
@@ -0,0 +1,369 @@
+  // but fresh identifiers we will use to map to the output of the corresponding
+  // child process.
+  // As such, let's just use the `pipe` setting for the stdio of this child process.
+  let output_fd = redirect[2];
+
+  /* In Javascript, there is no `execvp` and `execv` which
+     switch between searching in the PATH or not.
+     We will just run the vanilla `exec`, which should behave
+     as an `execvp`.
+     This might have slightly differing behavior than the
+     actual implementation.
+   */
+  if (optenv == 0) {
+    const child = globalThis.spawnSync(executable, argv, { stdio: "pipe" });
+
+    globalThis.pidToProcessTable.set(child.pid, child);
+    globalThis.fdTable.set(output_fd, child.stdout);
+
+    return child.pid;
+  } else {
+    let env = {};
+    for (const arg of optenv[1].slice(1)) {
+      const [key, value] = arg.split("=");
+      env[key] = value;
+    }
+    const child = globalThis.spawnSync(executable, argv, {
+      env: env,
+      stdio: "pipe",
+    });
+
+    globalThis.pidToProcessTable.set(child.pid, child);
+    globalThis.fdTable.set(output_fd, child.stdout);
+
+    let current_pid = globalThis.process.pid;
+
+    // associate parent to new child
+    if (globalThis.pidToChildrenTable.has(current_pid)) {
+      globalThis.pidToChildrenTable.set(
+        current_pid,
+        globalThis.pidToChildrenTable.get(current_pid).concat([child])
+      );
+    } else {
+      globalThis.pidToChildrenTable.set(current_pid, [child]);
+    }
+
+    return child.pid;
+  }
+}
+
+//Provides: unix_waitpid
+function unix_waitpid(flags, pid_req) {
+  console.error("unix_waitpid");
+  console.error(flags);
+  console.error(pid_req);
+
+  // TODO: flags stuff is not implemented
+  // TODO: WIFSTOPPED is not currently possible, because I don't know how to
+  // tell if a Javascript process is stopped or killed
+
+  function handleChild(child) {
+    // if it received a signal and is no longer running
+    // let's assume it was killed by the signal
+    let pid = child.pid;
+
+    if (child.status === null) {
+      // WSIGNALED of int
+      let status = [1, globalThis.os.constants.signals[child.signal]];
+      return [0, pid, status];
+    } else {
+      // WEXITED of int
+      let status = [0, child.status];
+      return [0, pid, status];
+    }
+  }
+
+  // any child process
+  if (pid_req == -1) {
+    let children = globalThis.pidToChildrenTable.get(process.pid);
+    if (!children) {
+      throw new Error(`unix_waitpid: no children process of parent {pid_req}`);
+    }
+    for (const child of children) {
+      return handleChild(child);
+    }
+    // specific process
+  } else if (pid_req > 0) {
+    let child = globalThis.pidToProcessTable.get(pid_req);
+    if (!child) {
+      throw new Error(`unix_waitpid: no child at pid {pid_req}`);
+    }
+
+    return handleChild(child);
+    // process group nonsense
+  } else {
+    function getAllDescendants(pid) {
```

Comment:
Why do we need to do this? Is this just future proofing or do will we actually hit this codepath w/ LSP.js?

Reply:
This is basically just because I wanted to accurately implement `unix_waitpid`, to a point. The process group stuff didn't seem hard so I decided to write it. I don't think anything exercises this at the moment.

Reply:
Fair, sounds good